openide.text/src/org/openide/text/UndoRedoManager.java

```diff
@@ -251,7 +251,7 @@ void beforeUndoAtSavepoint(WrapUndoEdit edit) {
     }
         
     private void undoSaveActions() {
-        if (onSaveTasksEdit != null) {
+        if (onSaveTasksEdit != null && onSaveTasksEdit.canUndo()){
```

Comment:
There's similar code in `redoSaveActions`; should it check the `canRedo()` as well ?

As I went through the relevant code In `UndoRedoManager`, there's one strange thing at line ~282 (in `afterUndoCheck`): when the edits are undone to the savepoint, the `onSaveTasksEdit` are '''redone''' (not undone). Compare with `afterRedoCheck`, actions are '''re'''done after redoing to the savepoint. This seems inconsistent; could you please check the behaviour ?

Reply:
redoSaveActions is called from delegateUndoFailedAtSavepoint/delegateRedoAtFailedAtSavepoint in finally block of undo/redo function of WrapUndoEdit class . It will call redoSaveActions only if delegate.undo() had thrown CannotUndoException in  WrapUndoEdit class.
As already it had thrown undoException, checking for onSaveTasksEdit.canUndo() inside undoSaveActions method will not make any difference. So should I check for canRedo() inside redoSaveActions?

"As I went through the relevant code In UndoRedoManager, there's one strange thing at line ~282 (in afterUndoCheck): when the edits are undone to the savepoint, the onSaveTasksEdit are '''redone''' (not undone)"
I had checked after changing onSaveTasksEdit.undo() to  onSaveTasksEdit.redo() at line 282; with refactor constructor call to factory method.
After refactoring  I did undo , redo and undo. 
Last undo actionthrows CannotUndoException at line 282 as onSaveTasksEdit.hasBeenDone flag is already true at that point.

Reply:
re: 
> So should I check for canRedo() inside redoSaveActions?
yes, whether to do it was the original question. 

Never mind, your subtle change is necessary as `undo` should be never called on an action that gives `!canUndo()`. Please do the same for the Redo, I don't know how to simulate situation which would trigger the bug, but the check follows the API spec. Accidentaly thrown `CannotUndoException` would certainly corrupt the undo/redo processing queue. 

Thanks.